### PR TITLE
Final PSF solution notebook

### PR DESCRIPTION
**Pull request by mcorrenti**

### CODE REVIEW

**Comment by ojustino** [comment]
Hi @mcorrenti,

Thank you for submitting these changes to your notebook. Please read on for your technical review.

## Before you begin

The technical review helps ensure that contributed notebooks a) run from top to bottom, b) follow the [PEP8 standards](https://www.python.org/dev/peps/pep-0008/) for Python code readability, and c) conform to the Institute's [style guide](https://github.com/spacetelescope/style-guides/blob/master/guides/jupyter-notebooks.md) for Jupyter Notebooks.

I've pushed the review as a new commit in this pull request. **To view and edit the commit locally, follow these steps:**

```
git checkout psf_final_nb
git fetch YOUR-REMOTE-FORK
git merge YOUR-REMOTE-FORK/psf_final_nb
```

_(`YOUR-REMOTE-FORK` is your fork's online copy. It's often `origin`, but if  you don't know your name for it, run `git remote -a` and choose the one whose path ends with `git@github.com:YOURUSERNAME/jwebbinar_prep.git`.)_

From here you can work on your branch as normal. If you have trouble with this step, please let me know before continuing.

---

## Instructions

After updating your local copy of this branch, **please open your notebook and address any warnings or errors you find**.

If you see cells with output like this, it means some of your code doesn't follow the PEP8 standards of code readability:

image

_(In the example above, `INFO - 3:3: E111` means that the text entered on line 3 at index 3 caused the warning "E111". The violation is briefly described at the end of the message.)_

You can test that your edits satisfy the standard by installing `flake8` on the command line with:
```
pip install flake8 pycodestyle_magic
```

Then, restart the notebook and run the following cells:

image

After that, edit and re-run cells with warnings until you've fixed all of them. **Please remember to delete the cells shown in the above image before pushing your changes back to this pull request.**

**To ask questions or leave feedback on specific cells, click the earlier message in this thread from the "review-notebook-app" bot.** There, you can comment on specific cells and view what's changed in the new commit. I may also write comments there, though all comments made on that page will also be reflected in this pull request's conversational thread.

---

## The three-point review (\*action required\*)

1. ⚠️ Notebook execution: **I was not able to run the notebook from top to bottom.**
    - Everything was going well until Section 9 ("Bonus part II"). Both of the `epsf_grid = build_epsf_grid()` cells returned an error and a blank plot, but I think it's due to an issue with arguments in that section's plotting functions. I will leave comments on the relevant cells.
    - In Section 10, `image_f200w`'s header doesn't seem to have the information that's being queried in the first few cells.

2. ⚠️ Code style: **There are a decent number of PEP8 violations.**
    - Please follow the advice in the "__Instructions__" section above to fix them.

3. ⚠️ Notebook style:
    - **I would either fill in or remove section 7.** I think it's appopriate to leave the interactive component to the "live" version and have this "solution" notebook be its fully realized counterpart.
    - I changed the section numbers in 5.5 and 5.6; they both originally both said 5.4.
    - I added a warning about the use of LaTeX in your `matplotlib` figures -- if the user doesn't have latex installed, they won't work.

File: pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb [hunk 1]
```diff
@@ -74,7 +74,7 @@
    "id": "95891849",
```
Comment (ojustino) [inline pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:74]: <pre class="ql-syntax" spellcheck="false">Line #3.    #os.environ['WEBBPSF_PATH'] = '/data/webbpsf-data'
</pre><p>&nbsp;This line should either be deleted or check whether the <code>$WEBBPSF_PATH</code> environment variable already exists.</p><p> </p><p>If you choose the latter, you could try something like:</p><pre class="ql-syntax" spellcheck="false">if not os.environ.get('WEBBPSF_PATH'):
    os.environ['WEBBPSF_PATH'] = '/data/webbpsf-data' 
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jwebbinar_prep/pull/66/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ojustino) [inline pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:74]: <p>Please uncomment or delete the commented lines of code in this cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jwebbinar_prep/pull/66/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (ojustino) [inline pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:74]: <p>The plot text functions you call in <code>plot_count_grid()</code> don't have <code>font</code> as an argument. Instead, try <code>fontsize=NUMBER</code> where you have <code>font={'size': NUMBER}</code>.<span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jwebbinar_prep/pull/66/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='111'/>

Comment (ojustino) [inline pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:74]: <p>The plot text functions you call in <code>plot_epsf()</code> don't have <code>font</code> as an argument. Instead, try <code>fontsize=NUMBER</code> where you have <code>font={'size': NUMBER}</code>.<span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jwebbinar_prep/pull/66/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='115'/>

Comment (ojustino) [inline pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:74]: <pre class="ql-syntax" spellcheck="false">Line #1.    from collections import OrderedDict
</pre><p>Please move this import to the top of the notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jwebbinar_prep/pull/66/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='116'/>

**Comment by ojustino** [comment]
Hello @mcorrenti, thank you for your revisions. I only have one remaining comment for you. If you decide to make further edits in response to it, please follow the instructions from the original review's "Before you begin" section before creating a new commit.

## The three-point review (\*action required\*)

1. ✅ Notebook execution: **I was able to run the notebook from top to bottom.**

2. ✅ Code style: **There are no remaining PEP8 violations.**
    - I adjusted a couple of line breaks to enhance readability. They don't affect the notebook's execution.

3. ⚠️ Notebook style:
    - The commented lines in `create_psf_model()` are still there. If they aren't needed for the function, it's better to delete them.

**Comment by mcorrenti** [comment]
@ojustino I would prefer to keep the commented lines in `create_psf_model()` so the users know how to modify the `psf_grid` function when they want to add the spectral source obtained using pysynphot.

File: pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb (hunk 1 above)
Comment (ojustino) [reply pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:73]: > @ojustino I would prefer to keep the commented lines in `create_psf_model()` so the users know how to modify the `psf_grid` function when they want to add the spectral source obtained using pysynphot.

In this case, I would add `source` as an argument of `create_psf_model()`.
1. I'd add `source=None` to the function definition. (`source` is eventually passed to WebbPSF's `calc_psf`, where `None` is the default value)
2. Inside `create_psf_model()`, keep the `psf = ...` lines that include `source` as an argument and delete the others.
3. Add a short docstring to the function that explains `source`, and you can move the sample call to `webbpsf.specFromSpectralType()` there.

Comment (ojustino) [reply pointsource_imaging/NIRCam_Basic_PSF_Photometry_solution.ipynb:73]: The latest commit implements this advice.

**Comment by ojustino** [comment]
Thanks for the quick update, @mcorrenti. You've now addressed all the concerns I raised.

@camipacifici We've successfully completed the technical review and are ready to merge this pull request.

**Comment by camipacifici** [comment]
Thank you both!

**Pull request merged**